site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md

```diff
@@ -0,0 +1,548 @@
+
+[octane]: https://emberjs.com
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+
+const MAX_LENGTH = 10;
+
+export default class PersonInfo extends Component {
+  @tracked name = '';
+
+  get nameLength() {
+    return this.name.length;
+  }
+
+  get remaining() {
+    return this.nameLength - MAX_LENGTH;
+  }
+
+  get showError() {
+    return this.remaining < 0;
+  }
+
+  updateNameTo = ({ target: { value } }) => this.myName = value;
+}
+```
+
+```handlebars
+<div>
+  <input {{on "input" this.updateName}} value={{this.myName}} />
+  <p class='{{if this.showError "error"}}'>
+    ({{this.nameLength}} remaining)
+  </p>
+</div>
+```
+
+There are a handful of interesting features to note about this code’s approach to reactivity. We decorate *one* piece of state, `name`, with `@tracked`, and the rest of the state updates automatically—including the `showError` and `remaining` properties don’t even refer to `name` directly. All of this with a particularly light touch:
+
+- There is no need to mark dependent keys on the getters (as in classic Ember components) and no need for a `computed` hash (as in Vue 2) for derived state: these are plain JavaScript getters.
+- There is no need for a dedicated utility like `setState` like in React’s class-based components or `set` from Ember Classic; this code just uses standard JavaScript assignment to update the value of `myName`.
+- This does not use two-way binding like *really old* Ember did or current day Angular or Vue do[^vue-2wb]—updates are explicit, but brief.
+
+This can look like magic when you first encounter it—especially the way undecorated getters update on demand. In fact, though, it’s *Just JavaScript™*, built on the combination of standard JavaScript patterns with a mix of computer science ideas ranging from tried-and-true ideas from decades ago to cutting-edge research. In the rest of this post, we’ll see how it works.
+
+[^vue-2wb]: Vue does not *require* two-way binding, but does make it *easy*.
+
+## How getters work
+
+First, let’s make sure we have a clear handle on how getters work in JavaScript in general. Once you understand this, seeing how autotracking works will be much easier. (If you already have a good understanding of the semantics and behavior of getters vs. assignment, feel free to [skip to the next section](#autotracking).) We’ll start by looking at the exact same class we started with, but with all of the Glimmer and DOM details removed, and a constructor added:
+
+```js
+const MAX_LENGTH = 10;
+
+export default class PersonInfo {
+  name;
+
+  constructor(name) {
+    this.name = name;
+  }
+
+  get nameLength() {
+    return this.name.length;
+  }
+
+  get remaining() {
+    return this.nameLength - MAX_LENGTH;
+  }
+
+  get showError() {
+    return this.remaining < 0;
+  }
+
+  updateNameTo = (value) => this.myName = value;
```

Comment:
I wonder if using arrow function syntax here might be confusing to some folks. I understand why you're doing it, to follow the Glimmer component example, but it's a bit tangential, and I don't think necessary for this particular example.

Reply:
On the flip side, explaining why you need to change it to be an arrow function later could also be tangential and confusing

Reply:
Yeah, I’ve gone back and forth on this. I think I’ll add a footnote calling it out. My thought was that I either need to do it this way *or* I need to explain binding with `@action` or via arrow function and it was easier to go this way given an expected/hoped-for broader audience than just Ember devs. :thinking: